packages/muelu/test/scaling/MatvecKernelDriver.cpp

```diff
@@ -100,6 +100,219 @@ void print_crs_graph(std::string name, const V1 rowptr, const V2 colind) {
+#include <cublas_v2.h>
+
+// The cuda examples provide a nice helper function that does all
+// of this error checking
+// helper_cuda.h
+// I am tempted to place it into our repo... but I'll need to read the license first.
+// Worst case, we could replicate here (the following is essentially part of the funcitonality
+// but they do it better.
+
+std::string cusparse_error(cusparseStatus_t code) {
+  switch(code) {
+  case CUSPARSE_STATUS_SUCCESS:
+    return std::string("CUSPARSE_STATUS_SUCCESS: Success");
+  case CUSPARSE_STATUS_NOT_INITIALIZED:
+    return std::string("CUSPARSE_STATUS_NOT_INITIALIZED: Library not initialized.");
+  case CUSPARSE_STATUS_ALLOC_FAILED:
+    return std::string("CUSPARSE_STATUS_ALLOC_FAILED: resources could not be allocated.");
+  case CUSPARSE_STATUS_INVALID_VALUE:
+    return std::string("CUSPARSE_STATUS_INVALID_VALUE: invalid parameters were passed (m,n,nnz<0).");
+  case CUSPARSE_STATUS_ARCH_MISMATCH:
+    return std::string("CUSPARSE_STATUS_ARCH_MISMATCH: the device does not support double precision (compute capability (c.c.) >= 1.3 required), symmetric/Hermitian matrix (c.c. >= 1.2 required), or transpose operation (c.c. >= 1.1 required).");
+  case CUSPARSE_STATUS_INTERNAL_ERROR:
+    return std::string("CUSPARSE_STATUS_INTERNAL_ERROR: Internal error");
+  case CUSPARSE_STATUS_MATRIX_TYPE_NOT_SUPPORTED:
+    return std::string("CUSPARSE_STATUS_MATRIX_TYPE_NOT_SUPPORTED: Matrix type not supported");
+  default:
+    return (std::string("ERROR: cuda_error got a return code it doesn't not understand") + std::to_string(code));
+  }
+}
+
+
+template<typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+class CuSparse_SpmV_Pack {
+  typedef double Scalar;
+  typedef Tpetra::CrsMatrix<Scalar,LocalOrdinal,GlobalOrdinal,Node> crs_matrix_type;
+  typedef Tpetra::MultiVector<Scalar,LocalOrdinal,GlobalOrdinal,Node> vector_type;
+  typedef typename crs_matrix_type::local_matrix_type    KCRS;
+  typedef typename KCRS::StaticCrsGraphType              graph_t;
+  typedef typename graph_t::row_map_type::non_const_type lno_view_t;
+  typedef typename graph_t::row_map_type::const_type     c_lno_view_t;
+  typedef typename graph_t::entries_type::non_const_type lno_nnz_view_t;
+  typedef typename graph_t::entries_type::const_type     c_lno_nnz_view_t;
+  typedef typename KCRS::values_type::non_const_type     scalar_view_t;
+  typedef typename Node::device_type device_type;
+
+  typedef typename Kokkos::View<int*,
+                                typename lno_nnz_view_t::array_layout,
+                                typename lno_nnz_view_t::device_type> cusparse_int_type;
+public:
+  CuSparse_SpmV_Pack (const crs_matrix_type& A,
+                      const vector_type& X,
+                            vector_type& Y)
+  {}
+
+ ~CuSparse_SpmV_Pack() {};
+
+  cusparseStatus_t spmv(const Scalar alpha, const Scalar beta) { return CUSPARSE_STATUS_SUCCESS; }
+};
+
+template<typename LocalOrdinal, typename GlobalOrdinal>
```

Comment:
If you're going to use Scalar instead of double throughout this class, why not just template the class on Scalar and do a partial specialization for `Scalar=double`?